### PR TITLE
Swift: resolve symlinks conditionally

### PR DESCRIPTION
**Pull request by AlexDenisov**

### CODE REVIEW

File: swift/extractor/infra/Path.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  }
+  return true;
+}
+
+std::filesystem::path getCodeQLPath(std::string_view path) {
+  // TODO: this needs more testing
+  // TODO: check canonicalization of names on a case insensitive filesystems
+  std::error_code ec;
+  std::filesystem::path ret = {};
+  auto canonicalize = shouldCanonicalize();
```
Comment (redsun82) [inline swift/extractor/infra/Path.cpp:24]: should we maybe make this `static const` so reading the environment is done exactly once?

**Review by redsun82 — COMMENTED** [review]
I'm not sure we can really commit symlinks to this repo to avoid problems checking this out on Windows, so maybe the symlinks need to be set up by `test.py` (and .gitignored)

File: swift/extractor/infra/Path.cpp (hunk 1 above)
Comment (AlexDenisov) [reply swift/extractor/infra/Path.cpp]: Done.

File: swift/integration-tests/posix-only/symlinks/test.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+from create_database_utils import *
+import os
+
+symlinks = ['preserve/Sources/A.swift', 'resolve/Sources/A.swift']
+
+for s in symlinks:
+    try:
+        os.symlink(os.getcwd() + '/main.swift', s)
```
Comment (redsun82) [inline swift/integration-tests/posix-only/symlinks/test.py:8]: Embrace `pathlib` 😜

```suggestion
        Path(s).symlink_to(Path('main.swift').resolve())
```

File: swift/integration-tests/posix-only/symlinks/test.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    except:
```
[the remark this answers is not in the document]
Comment (AlexDenisov) [reply swift/integration-tests/posix-only/symlinks/test.py:9]: There is always a room for improvement, but I decided to be rather concise here 😄

[the remark this answers is not in the document]
Comment (redsun82) [reply swift/integration-tests/posix-only/symlinks/test.py:9]:
```suggestion
    except FileExistsError:
```
does not seem to be too wordy :wink:

File: swift/integration-tests/posix-only/symlinks/test.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from create_database_utils import *
+import os
```
Comment (redsun82) [inline swift/integration-tests/posix-only/symlinks/test.py:2]:
```suggestion
from pathlib import Path
```
or my earlier proposed change won't work

**Comment by AlexDenisov** [comment]
@redsun82 I reverted the suggestion as they broke something once again.

**Comment by redsun82** [comment]
> @redsun82 I reverted the suggestions as they broke something once again.

sorry! I'm also ok with reverting and pushing as is, we can change this later, they were only nits either way

**Review by redsun82 — APPROVED** [review]
